### PR TITLE
PANIC if the workfile list in shared-memory is inconsistent

### PR DESCRIPTION
**Pull request by gfphoenix78**
`workfile_shared->num_active` may not the actual size of the list workfile_shared->activeList in production. The best way is to find and fix the inconsistency. However, we failed to find where the problem code is. The above two variables are both in shared memory. It's commonly to reset the share-memory when part of the shared memory is corrupted.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by gfphoenix78** [comment]
> Will you add the volatile keyword to protect register concurrent access? Such as workfile_shared->num_active.

It's not needed.

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 1]
```diff
@@ -563,6 +563,47 @@ workfile_mgr_close_set(workfile_set *work_set)
 	/* no op */
 }
 
+static void
+workfile_report_inconsistency()
```
Comment (pivotal-ning-yu) [inline src/backend/utils/workfile_manager/workfile_mgr.c:567]:
```suggestion
workfile_report_inconsistency(void)
```

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 2]
```diff
@@ -563,6 +563,47 @@ workfile_mgr_close_set(workfile_set *work_set)
...
+{
+	dlist_node *node;
+	dlist_iter		iter;
+	int num_freeList = 0;
+	int num_activeList = 0;
```
Comment (pivotal-ning-yu) [inline src/backend/utils/workfile_manager/workfile_mgr.c:572]:
```suggestion
	dlist_node *node;
	dlist_iter	iter;
	int			num_freeList = 0;
	int			num_activeList = 0;

```

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 3]
```diff
@@ -563,6 +563,47 @@ workfile_mgr_close_set(workfile_set *work_set)
...
+	dlist_foreach(iter, &workfile_shared->freeList)
+	{
+		num_freeList++;
+		node = iter.cur;
+		if (node->next == NULL || node->next->prev != node)
+			ereport(LOG, (errmsg("workfile freeList is corrupted")));
```
Comment (pivotal-ning-yu) [inline src/backend/utils/workfile_manager/workfile_mgr.c:578]: It's better to describe what's corrupted in the messages.

**Comment by gfphoenix78** [comment]
> BTW, I recall that @magi345 ever suggested marking the variable as `volatile`, that should be a necessary change, could we include it in this PR, too?

The similar code in twophase.c doesn't make the number volatile, it is accessed in the LWLock acquired section.

**Comment by magi345** [comment]
A simple check is to compile with/without volatile, and use `objdump -d` to get assemble code.  Compare the assembled code to see if any diff.
If volatile takes effect, the will be an extra `mov` instruction right before `add`

**Comment by gfphoenix78** [comment]
> A simple check is to compile with/without volatile, and use `objdump -d` to get assemble code. Compare the assembled code to see if any diff.
> If volatile takes effect, there will be an extra `mov` instruction right before `add`

Yes, the assembled code is different, but it doesn't necessarily affect the final result in this case.

**Comment by magi345** [comment]
As far as I know, LWLock and volatile are different things.
LWLock has a "Write barrier".

> Write barriers control the order of writes. Because writes to memory are slow (compared to the speed of the CPU), there is usually a write-request queue where writes are posted before they 'really happen'. Although they are queued in order, while inside the queue the writes may be reordered. (So maybe 'queue' isn't the best name...) Unless you use write barriers to prevent the reordering.

However, volatile behaves more like a "Read barrier"

> Read barriers control the order of reads. Because of speculative execution (CPU looks ahead and loads from memory early) and because of the existence of the write buffer (the CPU will read a value from the write buffer instead of memory if it is there - ie the CPU thinks it just wrote X = 5, then why read it back, just see that it is still waiting to become 5 in the write buffer) reads may happen out of order.

Refer: https://stackoverflow.com/questions/1787450/how-do-i-understand-read-memory-barriers-and-volatile

**Comment by gfphoenix78** [comment]
Typically,  the compiler will write back register values before a function call, thus ensuring consistency.

**Comment by adam8157** [comment]
I think I got this.

1, we don't need `volatile` in a critical section.
2, it's not about the function call, it's because of the barriers and the "acquire release" mechanism. PG made locks as compiler barriers and the compiler would be prevented from optimizing access to shared data within the critical section.

Too long, but please read:
"Spinlocks and compiler/memory barriers"
https://www.postgresql-archive.org/Spinlocks-and-compiler-memory-barriers-td5809455.html
"Why the “volatile” type class should not be used (put variables into a critical section instead)"
https://www.kernel.org/doc/html/latest/process/volatile-considered-harmful.html

File: src/backend/utils/workfile_manager/workfile_mgr.c (hunk 3 above)
Comment (gfphoenix78) [reply src/backend/utils/workfile_manager/workfile_mgr.c]: Updated

File: src/backend/utils/workfile_manager/workfile_mgr.c (hunk 1 above)
Comment (gfphoenix78) [reply src/backend/utils/workfile_manager/workfile_mgr.c]: Thanks, updated.

**Comment by magi345** [comment]
The metrics collector calls this function every few seconds on customer production env. Before this PR we only observed accidental memory corruption.
With this panic hammer, as long as inconsistency happens, the metrics collector will 100% crash at this function leads to GPDB down/recovery. The worst-case for the customer is service be suspended and a big business impact may happen.

So before shipping this to production release, I would like to make every effort to test it in our lab. Unfortunately, there is no integration test for the metrics collector and gpdb server.
My plan is to build a dev pipeline for this PR, bundle metrics collector to GPDB and test
- Run ICW
- Run TPCDS with periodical ANALYZEDB
I already set off the plan at https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/hw_6x_mc
@pivotal-ning-yu @gfphoenix78 Will you join me on building the test and wait more days until the test is done?

File: src/backend/utils/workfile_manager/workfile_mgr.c [hunk 4]
```diff
@@ -563,6 +563,55 @@ workfile_mgr_close_set(workfile_set *work_set)
...
+		if (node->next == NULL || node->next->prev != node)
+			ereport(LOG, (errmsg("workfile activeList is corrupted: "
+							"node = %p, next = %p, next->prev = %p",
+							node, node->next, node->next ? node->next->prev : NULL)));
+
+	}
+	if (!dlist_is_empty(&workfile_shared->activeList))
+	{
+		node = &workfile_shared->activeList.head;
+		if (node->next == NULL || node->next->prev != node);
```
Comment (d) [inline src/backend/utils/workfile_manager/workfile_mgr.c:603]: This trailing semicolon is sure not intended in this patch, no?

Comment (gfphoenix78) [reply src/backend/utils/workfile_manager/workfile_mgr.c:603]: OMG, thank you. It's not intended.

**Pull request merged**